src/main/java/bio/terra/workspace/common/utils/ControllerValidationUtils.java

```diff
@@ -42,24 +43,48 @@ public static void validatePaginationParams(int offset, int limit) {
   /**
    * Utility function for validating a CreateDataReferenceRequestBody.
    *
-   * <p>CreateDataReferenceRequestBody is currently structured to allow several parameters for
-   * controlled and private resources that aren't supported in WM. This function throws exceptions
-   * if any of those fields are set, or if any required fields are missing.
+   * <p>CreateDataReferenceRequestBody holds one of several types of data reference. This enforces
+   * that exactly one type of reference is present, and that it matches the specified ReferenceType
+   * field.
    */
   public static void validate(CreateDataReferenceRequestBody body) {
-    if (body.getResourceId() != null) {
-      throw new ControlledResourceNotImplementedException(
-          "Unable to create a reference with a resourceId, use a reference type and description"
-              + " instead. This functionality will be implemented in the future.");
+    ReferenceTypeEnum referenceType = body.getReferenceType();
+    // ReferenceType can be null if the user provides an invalid enum value.
+    if (referenceType == null) {
+      throw new ValidationException("Invalid reference type specified");
     }
-    if (body.getReferenceType() == null || body.getReference() == null) {
-      throw new InvalidDataReferenceException(
-          "Data reference must contain a reference type and a reference description");
+    // TODO(PF-404): requests using the deprecated reference field will not set the referenceInfo
+    // field. referenceInfo can be made required and this check can be removed once clients migrate.
+    DataReferenceInfo info =
+        body.getReferenceInfo() == null ? new DataReferenceInfo() : body.getReferenceInfo();
```

Comment:
This handles the existing case where a client specifies the old `reference` field but does not specify the new `referenceInfo` field. You can't have bucket or dataset references this way (as `getGoogleBucket()` and `getBigQueryDataset()` on this reference will be null), but you can specify a snapshot in the old `reference` field this way, and we want to support that until we move clients. That's why we check either the `referenceInfo` or the `reference` field when validating snapshots here.
(As above, existing clients do already use the `DATA_REPO_SNAPSHOT` referenceType, so that hasn't changed)